imgee/tasks.py

```diff
@@ -1,35 +1,68 @@
+import re
 import redis
 from imgee import app
 
 
 class TaskRegistry(object):
     def __init__(self, name='default', connection=None):
-        self.connection = redis.from_url(connection) if connection else None
+        self.connection = connection or self.set_connection_from_url()
         self.name = name
-        self.key = 'imgee:registry:%s' % name
+        self.key_prefix = 'imgee:registry:%s' % name
+        self.filename_pattern = '^[a-z0-9\_\.]+$'
+        self.pipe = self.connection.pipeline()
 
-    def set_connection(self, connection=None):
-        connection = connection or app.config.get('REDIS_URL')
-        self.connection = redis.from_url(connection)
+    def set_connection_from_url(self, url=None):
+        url = url or app.config.get('REDIS_URL')
+        self.connection = redis.from_url(url)
+        return self.connection
 
-    def add(self, taskid):
-        self.connection.sadd(self.key, taskid)
-
-    def remove(self, taskid):
-        self.connection.srem(self.key, taskid)
-
-    def remove_all(self):
-        for k in self.get_all_keys():
-            self.remove(k)
+    def key_for(self, taskid):
+        return u'{key_prefix}:{taskid}'.format(key_prefix=self.key_prefix, taskid=taskid)
 
     def __contains__(self, taskid):
-        return self.connection.sismember(self.key, taskid)
+        return len(self.connection.keys(self.key_for(taskid))) > 0
 
-    def keys_starting_with(self, query):
-        return filter(lambda k: k.startswith(query), self.connection.smembers(self.key))
+    def add(self, taskid, expire=60):
+        self.connection.set(self.key_for(taskid), taskid)
+        # setting TTL of 60 seconds for the key
+        # if the file doesn't get processed within 60 seconds,
+        # it'll be removed from the registry
+        self.connection.expire(self.key_for(taskid), expire)
 
     def search(self, query):
-        return filter(lambda k: str(query) in k, self.connection.smembers(self.key))
+        # >> KEYS imgee:registry:default:*query*
+        if not re.compile(self.filename_pattern).match(query):
```

Comment:
Also, abort with a custom exception instead of returning `list()`. This error should never occur as you're not passing in untrusted user input, so if it somehow does happen, you want it showing up in the server error log.